runtime/src/utils/buffer/pool.rs

```diff
@@ -0,0 +1,443 @@
+
+        bytes_to_copy
+    }
+
+    /// Put the given `page` into the buffer pool.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the provided page is not exactly PAGE_SIZE bytes long.
+    fn cache(&mut self, page: &mut [u8], blob_id: u32, page_num: u64) {
+        assert_eq!(page.len(), PAGE_SIZE);
+        if self.index.contains_key(&(blob_id, page_num)) {
+            // This can happen if different threads fault on the same page.
+            return;
+        }
+
+        let page_array = page.try_into().unwrap();
+
+        let key = (blob_id, page_num);
+        if self.cache.len() < self.capacity {
+            self.index.insert(key, self.cache.len());
+            self.cache.push(CacheEntry {
+                key,
+                referenced: AtomicBool::new(true),
+                data: Box::new(page_array),
+            });
+            return;
+        }
+
+        // Cache is full, find a page to evict.
+        while self.cache[self.clock].referenced.load(Ordering::Relaxed) {
+            self.cache[self.clock]
+                .referenced
+                .store(false, Ordering::Relaxed);
+            self.clock = (self.clock + 1) % self.cache.len();
+        }
+
+        // Evict the page by replacing it with the new page.
+        let entry = &mut self.cache[self.clock];
+        entry.referenced.store(true, Ordering::Relaxed);
+        assert!(self.index.remove(&entry.key).is_some());
+        self.index.insert(key, self.clock);
+        entry.key = key;
+        *entry.data = page_array;
+
+        // Move the clock forward.
+        self.clock = (self.clock + 1) % self.cache.len();
+    }
+
+    /// Read the specified bytes, preferentially from the buffer pool cache. Bytes not found in the
+    /// buffer pool will be read from the provided `blob` and cached for future reads.
+    async fn read<B: Blob>(
+        pool: Arc<RwLock<BufferPool<PAGE_SIZE>>>,
+        blob: &B,
+        blob_id: u32,
+        blob_len: u64,
+        mut buf: &mut [u8],
+        mut offset: u64,
+    ) -> Result<(), Error> {
+        // Read up to a page worth of data at a time from either the buffer pool or the underlying
+        // blob, until the requested data is fully read.
+        while !buf.is_empty() {
+            // Get a read lock on the buffer pool and see if we can get (some of) the data from it.
+            {
+                let buffer_pool = pool.read().await;
+                let count = buffer_pool.read_at(blob_id, buf, offset);
+                if count != 0 {
+                    offset += count as u64;
+                    buf = &mut buf[count..];
+                    continue;
+                }
+            }
+
+            // Page fault: fetch the page from the underlying blob since it wasn't in the buffer
+            // pool.
+            //
+            // Note that we hold no locks at this point, so it's possible multiple threads can fault
```

Comment:
This was the comment I brought up originally about the "stampeding herd" case (where `groupcache` is typically used in golang for slightly different tasks because it is aware of this: https://github.com/golang/groupcache).

Reply:
I've done some stuff related to this type of pattern in `broadcast` (where you more or less register a channel on the "fetching" instance that is closed when resolved).

Reply:
One way to handle this could just be having a map of locks per `blob_id` + `page_num`. That's probably pretty inefficient tho.